### PR TITLE
Implement Modbus UDP

### PR DESCRIPTION
**Pull request by StefanNienhuis**
This PR implements Modbus RTU over UDP.

Also implements handling of UDP addresses in Modbus CLI and documents this usage in the README.

This solves #50.

### CODE REVIEW

File: rtu_over_udp_client.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	// Set write and read timeout
+	// var timeout time.Time
+	// if mb.Timeout > 0 {
+	// 	timeout = mb.lastActivity.Add(mb.Timeout)
+	// }
+	// if err = mb.conn.SetDeadline(timeout); err != nil {
+	// 	return
+	// }
```
Comment (tretmar) [inline rtu_over_udp_client.go:57]: if not needed, you could even remove it

File: rtu_over_udp_client.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+	// Send the request
+	mb.logf("modbus: send % x\n", aduRequest)
+	if _, err = mb.conn.Write(aduRequest); err != nil {
+		return
+	}
+	function := aduRequest[1]
```
Comment (tretmar) [inline rtu_over_udp_client.go:64]: before accessing `aduRequest[1]`, it's safer to check if the length of `aduRequest` is >= 2.

File: rtu_over_udp_client.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	functionFail := aduRequest[1] & 0x80
+	bytesToRead := calculateResponseLength(aduRequest)
+
+	var n int
+	var n1 int
+	var data [rtuMaxSize]byte
+	//We first read the minimum length and then read either the full package
```
Comment (tretmar) [inline rtu_over_udp_client.go:71]:
```suggestion
	// We first read the minimum length and then read either the full package
```

File: rtu_over_udp_client.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	//or the error package, depending on the error status (byte 2 of the response)
```
Comment (tretmar) [inline rtu_over_udp_client.go:72]:
```suggestion
	// or the error package, depending on the error status (byte 2 of the response)
```

File: rtu_over_udp_client.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	n, err = io.ReadAtLeast(mb.conn, data[:], rtuMinSize)
+	if err != nil {
+		return
+	}
+	//if the function is correct
+	if data[1] == function {
```
Comment (tretmar) [inline rtu_over_udp_client.go:78]: adding a length check for the data slice here would also be safer

**Review by tretmar — COMMENTED** [review]
I am not deep into the UDP protocol itself and also not into the implementation here in the repo.
As we have no internal use of Modbus RTU via UDP, I also don't see an issue against adding it to the repo here.

I've added a few comments, nothing critical here

File: rtu_over_udp_client.go (hunk 3 above)
Comment (andig) [reply rtu_over_udp_client.go:71]: In general- please `gofmt` if not already done ;)

**Comment by StefanNienhuis** [comment]
@tretmar Thanks for the review. I've made the changes.

**Comment by andig** [comment]
What's missing to get this merged?

**Comment by andig** [comment]
@StefanNienhuis after looking through the code (and not being a network protocol guy) I have a situation here in X where we're sending UDP requests to a fixed port and the device replies on a fixed port:

```go
// receiver
addr, err := net.ResolveUDPAddr("udp", "0.0.0.0:8899")
if err != nil {
	return nil, err
}

instance.conn, err = net.ListenUDP("udp", addr)
if err != nil {
	return nil, err
}

// sender
addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(ip, "8899"))
if err == nil {
        // hand-built RTU
	_, err = m.conn.WriteToUDP([]byte{0xF7, 0x03, 0x89, 0x1C, 0x00, 0x7D, 0x7A, 0xE7}, addr)
}
```

Do you see any chance to support such a scenario? Or maybe I misunderstand this PR or UDP in general?

**Comment by StefanNienhuis** [comment]
@andig I’m not quite sure what your goal is, but I see it’s intended for Goodwe inverters, which is exactly what I built this for. The only thing that would be missing is supporting the AA55 prefix it sends before responses on some models, but that may not be required for your model.

**Comment by andig** [comment]
I think I'm confused by the fact that your code does not have a UDP listener, but I guess I'm just over-complicating UDP mentally.

**Comment by StefanNienhuis** [comment]
Go abstracts away large parts of the UDP protocol details in their network implementation. This allows you to have a UDP ‘connection’ that can send and receive data similar to TCP, even though there isn’t a persistent connection.

**Comment by tretmar** [comment]
Sorry, it is common for us that the authors themselves are merging the PR but I guess you either are not allowed or just not think about merging your own PRs 🤔 

As this has no impact for us, I will merge it now :)

**Comment by andig** [comment]
@StefanNienhuis confirmed working in https://github.com/evcc-io/evcc/discussions/12752#discussioncomment-9127102. Unfortunately we're now confronted with https://github.com/grid-x/modbus/pull/85#issuecomment-2052042978. Seems forking would be the right way to go to implement a workaround for this device defect.

**Comment by StefanNienhuis** [comment]
Yeah, although it's pretty trivial to implement, it may be out of scope for this project to implement such device specific behavior. Not sure who decides what goes in this project and what not, but since you merged this PR, what do you think @tretmar? 

In case you're not familiar, some families of Goodwe inverters accept their commands in normal Modbus format, but send their response in Modbus format with the bytes AA55 prefixed.

**Comment by tretmar** [comment]
I'd say that device specific behavior doesn't belong into this project here 🤔 

Forking brings the risk of getting out of sync really quickly. Another idea is to create an option that you can configure that can deal with the response prefix of AA55. And only use it for your project then.

**Comment by andig** [comment]
> Sorry, it is common for us that the authors themselves are merging the PR but I guess you either are not allowed or just not think about merging your own PRs 🤔

@tretmar nobody contributing here has that permission. See https://github.com/grid-x/modbus/issues/45#issuecomment-1002671524, three years old. There's still a queue of PRs that need attention like the important https://github.com/grid-x/modbus/pull/81.

> Forking brings the risk of getting out of sync really quickly. 

Not responding brings the risk of forking.

> Another idea is to create an option that you can configure that can deal with the response prefix of AA55. And only use it for your project then.

How would I do that without a fork or a merged PR?

**Comment by tretmar** [comment]
> How would I do that without a fork or a merged PR?

Well, either forking or merged PR ofc. Sure, if you face the issue that you have to wait multiple months to get it merged, I understand your concerns.

**Comment by RTTTC** [comment]
> Sorry, it is common for us that the authors themselves are merging the PR but I guess you either are not allowed or just not think about merging your own PRs 🤔

@dammarco - May I ask how have you became a contributor _with rights to merge_ in this project ?
I think it would beneficial for all of us if @andig got similar rights to merge here.
@gq0 - ?

**Comment by andig** [comment]
@dammarco works for gridx. No magic.

> I think it would beneficial for all of us if @andig got similar rights to merge here.

Its perfectly fine for maintainers to do that.

**Comment by andig** [comment]
@StefanNienhuis thank you for adding ModbusUDP. I've noticed that the current implementation does not use any timeouts. This can lead to indefinite wait times, at least in my tests. It would be great, if similar `timeout` to the TCP transporter could be implemented, both for reading and writing.

**Pull request merged**